crates/ruff/src/rules/refurb/rules/unnecessary_enumerate.rs

```diff
@@ -0,0 +1,311 @@
+use num_traits::ToPrimitive;
+use std::fmt;
+
+use crate::autofix::snippet::SourceCodeSnippet;
+use ruff_diagnostics::{AutofixKind, Diagnostic, Edit, Fix, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast as ast;
+use ruff_python_ast::{Arguments, Constant, Expr};
+use ruff_python_codegen::Generator;
+use ruff_python_semantic::helpers::is_unused;
+use ruff_text_size::{Ranged, TextRange};
+
+use crate::checkers::ast::Checker;
+use crate::registry::AsRule;
+
+/// ## What it does
+/// Checks for uses of `enumerate` that discard either the index or the value
+/// when iterating over a sequence.
+///
+/// ## Why is this bad?
+/// The built-in `enumerate` function is useful when you need both the index and
+/// value of a sequence.
+///
+/// If you only need the index or values of a sequence, you should iterate over
+/// `range(len(...))` or the sequence itself, respectively, instead. This is
+/// more efficient and communicates the intent of the code more clearly.
+///
+/// ## Example
+/// ```python
+/// for index, value in enumerate(sequence):
+///     print(index)
+///
+/// for index, value in enumerate(sequence):
+///     print(value)
+/// ```
+///
+/// Use instead:
+/// ```python
+/// for index in range(len(sequence)):
+///     print(index)
+///
+/// for value in sequence:
+///     print(value)
+/// ```
+///
+/// ## References
+/// - [Python documentation: `enumerate`](https://docs.python.org/3/library/functions.html#enumerate)
+#[violation]
+pub struct UnnecessaryEnumerate {
+    subset: EnumerateSubset,
+    iterable_suggestion: Option<SourceCodeSnippet>,
+}
+
+impl Violation for UnnecessaryEnumerate {
+    const AUTOFIX: AutofixKind = AutofixKind::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let UnnecessaryEnumerate {
+            subset,
+            iterable_suggestion,
+        } = self;
+        if let Some(suggestion) = iterable_suggestion
+            .as_ref()
+            .and_then(SourceCodeSnippet::full_display)
```

Comment:
You could probably just use `map` instead:
```suggestion
        if let Some(suggestion) = iterable_suggestion.map(SourceCodeSnippet::full_display)
```

Reply:
The suggestion causes a type mismatch, is there a better way?